### DISCUSSION TITLE
How well is persistence in wasm supported?

### DISCUSSION THREAD

**Discussion by DAT4** [fyne-io/fyne#3407] [Q&A]
I am implementing an application now and testing it on linux, android and firefox. (it is probably mostly gonna be used in browsers)

The application requires login and is then using a token to call a websocket api to stream some commands which are used for opening a door.

So far I had a lot of issues with wasm because of stuff related to websocket :-D But I made it work perfectly now. 

There is just one feature I would like to implement: which is persistence of the token so that the user wont have to login every time.

I am wondering if the preference manager in fyne supports webassembly or if I have to implement the persistence functionality myself with the js syscalls?

**Comment by andydotxyz** [comment]
Unfortunately we don't yet have Preferences working with WASM/WebJS - we intend to do this through local storage, but it is not yet implemented.
Apologies for the inconvenience.

**Reply by DAT4** [reply]
@andydotxyz  Thank you for the quick reply - No worries I will make my own implementation for now 

When you are looking into implementing persistence for WASM check out OPFS - I think it is more secure than local storage 

https://sqlite.org/wasm/doc/tip/persistence.md 
https://webkit.org/blog/12257/the-file-system-access-api-with-origin-private-file-system/

**Comment by DAT4** [comment]
Implementation with localstorage was quite simple. 

I just made  a localStorage object which is then passed into the part of the program that needs to handle that.

```go
type LocalStorage struct {
}

func (s LocalStorage) GetToken() string {
	result := js.Global().Get("window").Get("localStorage").Call("getItem", "token")

	if result.IsNull() {
		return ""
	}

	return result.String()
}

func (s LocalStorage) SetToken(token string) {
	js.Global().Get("window").Get("localStorage").Call("setItem", "token", token)
}

```

**Reply by Bluebugs** [reply]
Nice, thanks for figuring it out. We can't add the support in this coming release as we are in feature freeze, but will make it happen for next one.
